### PR TITLE
feat: add dependency ID to az key vault dependency tracking

### PR DESCRIPTION
**Pull request by stijnmoreels**
Add dependency ID method overload for tracking Azure Key Vault dependencies.

Relates to #243

### CODE REVIEW

File: src/Arcus.Observability.Telemetry.Serilog.Sinks.ApplicationInsights/Converters/DependencyTelemetryConverter.cs [hunk 1]
```diff
@@ -34,12 +34,13 @@ protected override DependencyTelemetry CreateTelemetryEntry(LogEvent logEvent, I
...
-            
+
+            string dependencyId = logEntry.Properties.GetAsRawString(ContextProperties.DependencyTracking.DependencyId);
             string operationId = logEvent.Properties.GetAsRawString(ContextProperties.Correlation.OperationId);
-            
+
             var dependencyTelemetry = new DependencyTelemetry(dependencyType, target, dependencyName, data, startTime, duration, resultCode, success: outcome)
             {
-                Id = operationId
+                Id = dependencyId ?? operationId
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Telemetry.Serilog.Sinks.ApplicationInsights/Converters/DependencyTelemetryConverter.cs:43]: Not sure if this is still correct, we shouldn't use the operation ID anymore, right?

File: docs/preview/02-Features/writing-different-telemetry-types.md [hunk 2]
```diff
@@ -388,6 +388,19 @@ catch (Exception exception)
 }
 ```
 
+### Making it easier to link dependencies
+
+Service-to-service correlation requires linkage between tracked dependencies (outgoing) and requests (incoming).
+Tracking any kind of dependency with the library has the possibility to provide an dependency ID. This ID will be needed later when the incoming request is tracked (dependency ID = request's parent ID).
+
+// TODO: provide example.
```
Comment (fgheysels) [inline docs/preview/02-Features/writing-different-telemetry-types.md:396]: Think the example still needs to be done :)

File: src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs [hunk 3]
```diff
@@ -493,7 +493,84 @@ public static class ILoggerExtensions
...
+            Guard.NotNull(logger, nameof(logger), "Requires an logger instance to write the Azure Key Vault dependency");
+            Guard.NotNullOrWhitespace(vaultUri, nameof(vaultUri), "Requires a non-blank URI for the Azure Key Vault");
+            Guard.NotNullOrWhitespace(secretName, nameof(secretName), "Requires a non-blank secret name for the Azure Key Vault");
+            Guard.NotNull(measurement, nameof(measurement), "Requires a dependency measurement instance to track the latency of the Azure Key Vault when tracking an Azure Key Vault dependency");
+            Guard.For<FormatException>(
+                () => !SecretNameRegex.IsMatch(secretName), 
+                "Requires a Azure Key Vault secret name in the correct format, see https://docs.microsoft.com/en-us/azure/key-vault/general/about-keys-secrets-certificates#objects-identifiers-and-versioning");
+            Guard.For<UriFormatException>(
+                () => !KeyVaultUriRegex.IsMatch(vaultUri),
+                "Requires the Azure Key Vault host to be in the right format, see https://docs.microsoft.com/en-us/azure/key-vault/general/about-keys-secrets-certificates#objects-identifiers-and-versioning");
```
Comment (fgheysels) [inline src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs:531]: Where are these regexes defined ?
I'm just wondering if this kind of validation should be the responsability of Arcus ?  What if the format rules of secrentmaes are being changed ?  What if the documentation page is moved to another location ?

I think it is good to point people to the right direction, but this also puts an additional responsability to the Arcus lib as well

Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs:531]: Feel free to create an issue for this 👍

File: docs/preview/02-Features/writing-different-telemetry-types.md (hunk 2 above)
Comment (stijnmoreels) [reply docs/preview/02-Features/writing-different-telemetry-types.md]: Ok, I updated this, but we also should probably create a dedicated feature doc page to give a rundown of the entire service-to-service functionality. But that's for the future 😉 . I'll create an issue for this.

Comment (stijnmoreels) [reply docs/preview/02-Features/writing-different-telemetry-types.md]: #258

File: src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs (hunk 3 above)
Comment (fgheysels) [reply src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs:531]: I think a discussion might be more appropriate:
https://github.com/arcus-azure/arcus.observability/discussions/260

File: src/Arcus.Observability.Telemetry.Serilog.Sinks.ApplicationInsights/Converters/DependencyTelemetryConverter.cs (hunk 1 above)
Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Serilog.Sinks.ApplicationInsights/Converters/DependencyTelemetryConverter.cs:43]: Yes, let's remove this, as the `OperationId` is already assigned in the operation context information of the telemetry.

File: docs/preview/02-Features/writing-different-telemetry-types.md [hunk 4]
```diff
@@ -388,6 +389,33 @@ catch (Exception exception)
 }
 ```
 
+### Making it easier to link services
+
+Service-to-service correlation requires linkage between tracked dependencies (outgoing) and requests (incoming).
+Tracking any kind of dependency with the library has the possibility to provide an dependency ID. This ID will be needed later when the incoming request is tracked (dependency ID = request's parent ID).
```
Comment (fgheysels) [inline docs/preview/02-Features/writing-different-telemetry-types.md:395]:
```suggestion
Tracking any kind of dependency with the library has the possibility to provide a dependency ID. This dependency ID will be needed later when the incoming request is tracked (dependency ID = request's parent ID).
```

Comment (fgheysels) [reply docs/preview/02-Features/writing-different-telemetry-types.md:395]: I'm not sure how to interpret the last sentence ? The (dependency) ID is needed later when the incoming request is tracked.   You mean that the dependencyId is used in the logging of the downstream dependency to correlate the request id with the dependency id ?

**Comment by fgheysels** [comment]
I think some things need to be clarified in the docs before we can merge this.

File: docs/preview/02-Features/writing-different-telemetry-types.md (hunk 4 above)
Comment (stijnmoreels) [reply docs/preview/02-Features/writing-different-telemetry-types.md:395]: Yes, something like that. Maybe we should specify that the 'later' part is in the other dependency.

**Comment by stijnmoreels** [comment]
> I think some things need to be clarified in the docs before we can merge this.

Besides the open comment, something else?

File: docs/preview/02-Features/writing-different-telemetry-types.md (hunk 4 above)
Comment (stijnmoreels) [reply docs/preview/02-Features/writing-different-telemetry-types.md]: I updated the last sentence with a more thorough explanation. We also need to update the links but that can only be done when #250 is done.

File: src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs [hunk 5]
```diff
@@ -493,7 +493,84 @@ public static class ILoggerExtensions
...
+            TimeSpan duration,
+            Dictionary<string, object> context = null)
+        {
+            Guard.NotNull(logger, nameof(logger), "Requires an logger instance to write the Azure Key Vault dependency");
+            Guard.NotNullOrWhitespace(vaultUri, nameof(vaultUri), "Requires a non-blank URI for the Azure Key Vault");
+            Guard.NotNullOrWhitespace(secretName, nameof(secretName), "Requires a non-blank secret name for the Azure Key Vault");
+            Guard.For<FormatException>(
+                () => !SecretNameRegex.IsMatch(secretName), 
+                "Requires a Azure Key Vault secret name in the correct format, see https://docs.microsoft.com/en-us/azure/key-vault/general/about-keys-secrets-certificates#objects-identifiers-and-versioning");
+            Guard.For<UriFormatException>(
```
Comment (fgheysels) [inline src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs:567]: I'd say we remove the KeyVault - name related validations as discussed in this discussion: #260

**Review by fgheysels — APPROVED** [review]
Looks good to me, but I think I'd remove the secretname & keyvaulturi validation

**Pull request merged**